File: 백준(최대한 많은 문제를 접해보자!)/400-다이나믹프로그래밍1/13 제곱수합.js
const fs = require('fs');
const input = fs.readFileSync('dev/stdin').toString().split('\n');

let N = Number(input[0]);
let dp = new Array(N + 1).fill(Number.MAX_SAFE_INTEGER);

dp[1] = 1;
let powArr = [1];
for (let i = 2; i <= N; i++) {
  // 1. 제곱 수로 이루어지는 수인지 확인
  let sqrtNum = Math.sqrt(i);
  let floorNum = Math.floor(sqrtNum);
  if (floorNum * floorNum === i) {
    dp[i] = 1;
    powArr.push(floorNum * floorNum);
    continue;
  }

  // 2. 제곱 수들을 이용하여 찾는다.
  for (let j = 0; j < powArr.length; j++) {
    dp[i] = Math.min(dp[i], 1 + dp[i - powArr[j]]);
  }
}
console.log(dp[N]);

//
//
//

const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = Number(fs.readFileSync(filePath).toString());
console.log(solution(input));

/**
 * @param {number} input
 */
function solution(input) {
  const cache = new Array(input + 1).fill(0).map((_, index) => index);

  cache[1] = 1;
  for (let i = 1; i <= input; i++) {
    for (let j = 1; j ** 2 <= i; j++) {
      cache[i] = Math.min(cache[i], cache[i - j ** 2] + 1);
    }
  }
  return cache[input];
}

File: 백준(최대한 많은 문제를 접해보자!)/400-다이나믹프로그래밍1/13 제곱수합.py
N = int(input())
dp = [0] * (N+1)
for i in range(1, N+1):
    dp[i] = i
    for j in range(1, int(i**0.5)+1):
        dp[i] = min(dp[i], dp[i-j**2]+1)
print(dp[N])

##
##
## 다른 풀이
n = int(input())
 
dp = [i for i in range (n+1)]
 
for i in range(1, n+1):
    for j in range(1, i):
        if (j * j) > i:
            break
        if dp[i] > dp[i - j * j] + 1:
            dp[i] = dp[i - j * j] + 1
 
print(dp[n])      